src/Datadog.Trace.ClrProfiler.Managed/AutoInstrumentation/CosmosDb/QueryIteratorsIntegrations.cs

```diff
@@ -0,0 +1,308 @@
+        ParameterTypeNames = new[] { MicrosoftAzureCosmosQueryDefinitionTypeName, ClrNames.String, MicrosoftAzureCosmosQueryRequestOptionsTypeName, },
+        MinimumVersion = Major3Minor0,
+        MaximumVersion = Major3MinorX,
+        IntegrationName = IntegrationName)]
+    [InstrumentMethod(
+        AssemblyName = MicrosoftAzureCosmosClientAssemblyName,
+        TypeName = "Microsoft.Azure.Cosmos.DatabaseCore",
+        MethodName = "GetUserQueryIterator",
+        ReturnTypeName = MicrosoftAzureCosmosFeedIteratorTypeName,
+        ParameterTypeNames = new[] { ClrNames.String, ClrNames.String, MicrosoftAzureCosmosQueryRequestOptionsTypeName, },
+        MinimumVersion = Major3Minor0,
+        MaximumVersion = Major3MinorX,
+        IntegrationName = IntegrationName)]
+
+    // Database level instrumentations for quering containers
+    [InstrumentMethod(
+        AssemblyName = MicrosoftAzureCosmosClientAssemblyName,
+        TypeName = "Microsoft.Azure.Cosmos.CosmosClient",
+        MethodName = "GetDatabaseQueryIterator",
+        ReturnTypeName = MicrosoftAzureCosmosFeedIteratorTypeName,
+        ParameterTypeNames = new[] { MicrosoftAzureCosmosQueryDefinitionTypeName, ClrNames.String, MicrosoftAzureCosmosQueryRequestOptionsTypeName, },
+        MinimumVersion = Major3Minor0,
+        MaximumVersion = Major3MinorX,
+        IntegrationName = IntegrationName)]
+    [InstrumentMethod(
+        AssemblyName = MicrosoftAzureCosmosClientAssemblyName,
+        TypeName = "Microsoft.Azure.Cosmos.CosmosClient",
+        MethodName = "GetDatabaseQueryIterator",
+        ReturnTypeName = MicrosoftAzureCosmosFeedIteratorTypeName,
+        ParameterTypeNames = new[] { ClrNames.String, ClrNames.String, MicrosoftAzureCosmosQueryRequestOptionsTypeName, },
+        MinimumVersion = Major3Minor0,
+        MaximumVersion = Major3MinorX,
+        IntegrationName = IntegrationName)]
+    [InstrumentMethod(
+        AssemblyName = MicrosoftAzureCosmosClientAssemblyName,
+        TypeName = "Microsoft.Azure.Cosmos.CosmosClient",
+        MethodName = "GetDatabaseQueryStreamIterator",
+        ReturnTypeName = MicrosoftAzureCosmosFeedIteratorTypeName,
+        ParameterTypeNames = new[] { MicrosoftAzureCosmosQueryDefinitionTypeName, ClrNames.String, MicrosoftAzureCosmosQueryRequestOptionsTypeName, },
+        MinimumVersion = Major3Minor0,
+        MaximumVersion = Major3MinorX,
+        IntegrationName = IntegrationName)]
+    [InstrumentMethod(
+        AssemblyName = MicrosoftAzureCosmosClientAssemblyName,
+        TypeName = "Microsoft.Azure.Cosmos.CosmosClient",
+        MethodName = "GetDatabaseQueryStreamIterator",
+        ReturnTypeName = MicrosoftAzureCosmosFeedIteratorTypeName,
+        ParameterTypeNames = new[] { ClrNames.String, ClrNames.String, MicrosoftAzureCosmosQueryRequestOptionsTypeName, },
+        MinimumVersion = Major3Minor0,
+        MaximumVersion = Major3MinorX,
+        IntegrationName = IntegrationName)]
```

Comment:
I wonder if it would make more sense to split some of these into separate integrations rather than reusing the same class. With the current design we have to do a lot of `if...else if` and trying duck casts. If they were split to support just the expected parameters we could avoid that complexity + overhead? Most of the logic could move to a Helper class that does the actual span creation, similar to some of the other integrations

Reply:
I've split this into a common class which is then called from classes that are specific to the type of integration, this is similar to the way GraphQL and a few other integrations are implemented. 

Using methods specific to each method signature and `TryDuckCast` reduces nesting.